src/models/ThemeProvider.js

```diff
@@ -1,121 +1,38 @@
-|}
-
-let warnChannelDeprecated
-if (process.env.NODE_ENV !== 'production') {
-  warnChannelDeprecated = once(() => {
-    // eslint-disable-next-line no-console
-    console.error(
-      `Warning: Usage of \`context.${CHANNEL}\` as a function is deprecated. It will be replaced with the object on \`.context.${CHANNEL_NEXT}\` in a future version.`
-    )
-  })
 }
 
 const isFunction = test => typeof test === 'function'
 
-/**
- * Provide a theme to an entire react component tree via context and event listeners (have to do
- * both context and event emitter as pure components block context updates)
- */
-export default class ThemeProvider extends Component<ThemeProviderProps, void> {
-  broadcast: Broadcast
-  getTheme: (theme?: Theme | ((outerTheme: Theme) => void)) => Theme
-  outerTheme: Theme
-  props: ThemeProviderProps
-  unsubscribeToOuterId: number = -1
-  unsubscribeToOuterId: string
-
-  static childContextTypes = contextShape
-  static contextTypes = {
-    [CHANNEL_NEXT]: CONTEXT_CHANNEL_SHAPE,
-  }
-
-  constructor() {
-    super()
-    this.getTheme = this.getTheme.bind(this)
-  }
+const ThemeContext = createContext()
 
-  componentWillMount() {
-    // If there is a ThemeProvider wrapper anywhere around this theme provider, merge this theme
-    // with the outer theme
-    const outerContext = this.context[CHANNEL_NEXT]
-    if (outerContext !== undefined) {
-      this.unsubscribeToOuterId = outerContext.subscribe(theme => {
-        this.outerTheme = theme
-
-        if (this.broadcast !== undefined) {
-          this.publish(this.props.theme)
-        }
-      })
-    }
-
-    this.broadcast = createBroadcast(this.getTheme())
-  }
+export const ThemeConsumer = ThemeContext.Consumer
 
-  getChildContext() {
-    return {
-      ...this.context,
-      [CHANNEL_NEXT]: {
-        getTheme: this.getTheme,
-        subscribe: this.broadcast.subscribe,
-        unsubscribe: this.broadcast.unsubscribe,
-      },
-      [CHANNEL]: subscriber => {
-        if (process.env.NODE_ENV !== 'production') {
-          warnChannelDeprecated()
-        }
-
-        // Patch the old `subscribe` provide via `CHANNEL` for older clients.
-        const unsubscribeId = this.broadcast.subscribe(subscriber)
-        return () => this.broadcast.unsubscribe(unsubscribeId)
-      },
-    }
-  }
-
-  componentWillReceiveProps(nextProps: ThemeProviderProps) {
-    if (this.props.theme !== nextProps.theme) {
-      this.publish(nextProps.theme)
-    }
-  }
-
-  componentWillUnmount() {
-    if (this.unsubscribeToOuterId !== -1) {
-      this.context[CHANNEL_NEXT].unsubscribe(this.unsubscribeToOuterId)
-    }
+/**
+ * Provide a theme to an entire react component tree via context
+ */
+export default class ThemeProvider extends Component<Props> {
+  getContext: (
+    theme: Theme | ((outerTheme: Theme) => void),
+    outerTheme?: Theme
+  ) => Theme
+
+  constructor(props: Props) {
+    super(props)
+    this.getContext = memoize(this.getContext.bind(this))
```

Comment:
Don't need to do `constructor` if you use this pattern on the class itself:

```js
class ThemeProvider extends Component<Props> {
  getContext = memoize(
    (theme: (outerTheme: ?Theme) => void, outerTheme?: Theme) => this.getTheme(theme, outerTheme)
  )
}
```

Reply:
Yeah I actually wrote it similar to that first, but refactored after reading this [thread](https://github.com/facebook/react/issues/9851). I was worried about booth perf and that the output isn't the same. Example [here](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=PTAEBEFMGMBsEMBOlT1AW0gFwBYHsATVAZ1QDtQBLM4reM6FAB0TyckSwE8AoOeYqQBiePKADePUKABmo0AF5QACgCUigHyhoeGnliQAdLDwBzZQHI5eC6p4BfHk5CgAQtQLVTqDNnxEBcioaOgZmVnZOXn5BNyQJKW1dWkQAV2gsPEQ1BOlpXEpiQwAjeKUCotLEEo9lCtUAbkTHaUSqnMk8pL0DYzNLKtsm6Ud7IA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=6.26.0&envVersion=).

However if you feel that that's not a problem I'll happily change

Reply:
We use loose mode, so ends up looking more like this: https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=true&code_lz=PTAEBEFMGMBsEMBOlT1AW0gFwBYHsATVAZ1QDtQBLM4reM6FAB0TyckSwE8AoOeYqQBiePKADePUKABmo0AF5QACgCUigHyhoeGnliQAdLDwBzZQHI5eC6p4BfHk5CgAQtQLVTqDNnxEBcioaOgZmVnZOXn5BNyQJKW1dWkQAV2gsPEQ1BOlpXEpiQwAjeKUCotLEEo9lCtUAbkTHaUSqnMk8pL0DYzNLKtsm6Ud7IA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&sourceType=module&lineWrap=true&presets=react%2Cstage-2%2Cenv&prettier=false&targets=&version=6.26.0&envVersion=1.6.2

Reply:
You can leave it though, I'll experiment with it both ways later